Use urwid Button features to simplify TopButton.

Rohitth007 opened the pull request.

neiljp commented on zulipterminal/ui_tools/buttons.py:
```diff
@@ -31,28 +31,31 @@ def __init__(
-        self.controller = controller
         urwid.connect_signal(self, "click", self.activate)
 
+        self.button_prefix = urwid.Text("")
+        self._w.contents[0] = (self.button_prefix, self._w.options("pack"))
+
+        self._label.set_wrap_mode("ellipsis")
+        self._label._cursor_position = 80
+        self._w.dividechars = 0
+
+        self.button_suffix = urwid.Text("")
+        self._w.contents[2] = (self.button_suffix, self._w.options("pack"))
+
+        self._w = urwid.AttrMap(self._w, None, "selected")
```
Regarding the use of private attributes, I recognize that we already reset the internal widget in the existing implementation, and so it definitely seems less of an issue to set `self._w` once (since we do already) - and certainly an improvement to set up the widgets once here rather than on every update, since we can with this kind of approach.

I also recognize the benefit of reusing `self._label` to allow the reuse in the update method more easily and potentially allow setting the label elsewhere - eg. if we update stream, topic or user text.

What I'm less clear of is if it would be clearer to construct our own Columns widget in an AttrMap (using self._label), so we get precisely what we want more explicitly.

Rohitth007 replied:
Yes agreed, creating our own widget once and replacing _w once would also be cleaner (contents list would not be needed).

neiljp reviewed (commented):
@Rohitth007 I just gave this a quick test and it seems behaviorally fine as a refactor except for the difference in spacing behavior I noted inline (that could give us an extra character, but spills into the unreads column just a bit too much, style-wise?)

This definitely has potential to simplify things - we spent quite some time on the ellipsis code in January-July 2019 or earlier, only for the feature to be implemented in urwid in probably 2.1.0 in November 2019.

As inline, I like a lot of this, my only query is how important it is to reuse the existing data-structure (self._*) vs be more explicit in creating what we need.

Rohitth007 commented on zulipterminal/ui_tools/buttons.py:
```diff
@@ -67,33 +70,17 @@ def update_count(self, count: int, text_color: Optional[str] = None) -> None:
+        if count_text[1]:
+            suffix = [" ", count_text, " "]
+        else:
+            suffix = " "
```
But why do we need 2 spaces at the end?

neiljp replied:
This reproduces the current behavior - I think otherwise it can look a little strange with the text between all of the unread counts, though I'm not sure if the current behavior was intentional. We can see what others think, though this won't be a pure refactor if we change the appearance.

Rohitth007 commented:
@neiljp I've updated the PR incorporating the suggestions. I've also added a new commit at the end dealing with a few more simplifications. I have not added the tests though. So only the first 3 pass.

neiljp commented on tests/ui_tools/test_buttons.py:
```diff
@@ -23,201 +22,120 @@
+        assert top_button.button_label.get_cursor_coords("size") is None
+        # assert top_button.button_suffix == self.urwid.Text("")
+
+        self.urwid.Columns.assert_called_once_with(
+            [
+                ("pack", top_button.button_prefix),
+                top_button.button_label,
+                ("pack", top_button.button_suffix),
+            ]
+        )
```
I know we have some of these kinds of asserts in other tests, but it'd be good to avoid them if possible, as they basically just mimic the implementation.

I'm wondering if it would be best to replace these kinds of lines with something that compares a render output.

Rohitth007 replied:
But that way we can only test the actual text we are assigning it ( "" for example ) not the actual thing we see (with spacing) as that is handled internally by urwid and depends on width.
Another thing is that this way we can also check the style we are assigning like "selected" for AttrMap.
There is no proper "output" here that we can check against.

neiljp reviewed (commented):
@Rohitth007 We discussed this in the stream. I think it will be less confusing to use the super() init, reuse the label and set the new widget - with a comment as to why we're doing so. The existing code is not clear and this will help a lot.

I'm not sure if we strictly need to reorder the parameters we're setting, but we could choose to squash before merging later.

I've not looked into everything here, but removing the width parameter looks like a great improvement and the logical follow-up - I didn't like having that present when it was implemented, but it seemed necessary at the time.

Rohitth007 commented on zulipterminal/ui_tools/buttons.py:
```diff
@@ -31,29 +31,32 @@ def __init__(
+            ]
+        )
+        self._w = urwid.AttrMap(cols, None, "selected")
+        urwid.connect_signal(self, "click", self.activate)
```
Usually this this is done after the Button widget is created, so I thought it should go after `_w`

Rohitth007 commented on zulipterminal/ui_tools/buttons.py:
```diff
@@ -68,33 +71,17 @@ def update_count(self, count: int, text_color: Optional[str] = None) -> None:
+            suffix = ["  "]
+        self.button_prefix.set_text(prefix)
+        self.button_label.set_text((text_color, self._caption))
+        self.button_suffix.set_text(suffix)
```
Do you mean like `self.button_prefix.set_text([" ", self.prefix_character, " "])`?
This causes 2 problems:
* If prefix_character is "" then it breaks the linebox (like emoji)
* There would be 2 spaces at the start

neiljp replied:
Let's discuss in the stream - but the idea would be to place the always-present left and right padding spaces in the later part.

Rohitth007 commented:
I have updated this PR y using `_label` and also including tests for the last commit.

neiljp commented:
@Rohitth007 I really wanted to get this tidy in - it reduces a lot of boilerplate and we can build upon it later :+1:
I've made a few minor changes and pushed back here, without the last commit since that seems separate and in-progress on top - and I'm merging now! :tada:

The tests were initially somewhat of a stumbling block to review due to the diff offsets as they were, so I extracted the derived class text_content tests, which were for the most part duplicates and barely changed the amount of coverage - these are already merged separately. I also reordered the connect_signal to sit at the end of the TopButton to simplify that diff and retain the original order, which is the only net change of code from your first 3 commits. The diffs are simpler now.

Your commit titles are better here, but note that your original titles were not unique in explaining eg. which init you were adjusting from reading just the title. Its also preferable to not refer to relative commits and be more descriptive (ie. removing the widths), though we do sometimes refer to (merged) commit refs, mainly in the commit text.

Now that this is merged, possible follow-up ideas:
* iterating on this design (the spaces idea I mentioned in the comments)
* tidying up that original last commit - as in other cases, seeing it as multiple commits may make it clearer what each change of code is
* possibly investigating using `.render()`, or probing the contents of structures, instead of `assert_called*` in UI tests?
